### ISSUE TITLE
[REQUEST] Function to register a callback on a successful partial match

### ISSUE THREAD

**Issue by mmccool** [ajv-validator/ajv#902]
# Change proposal
I would like the ability to register a callback function that gets called whenever there is a successful partial match of a specific part of the schema to the data.   This would look something like this:
```js
ajv.on("JSON Pointer into Schema",function (schema, data, other...) {....});
```

### The problem: 
We are working on a W3C standard (see https://github.com/w3c/wot-thing-description) that requires an "implementation report" (see testing/report.html in that repo for a draft) that tallys what features are being used or not by a large number of JSON files.  Part of what makes this complicated is the fact that JSON Schemas are not official standards yet so we can't just provide the schema as a specification, but have to relate it to a bunch of prose assertions in the standard itself.   Each assertion defines a feature which is tested by some part of the schema.  What we want to do is match all our test files against a single (complex) schema and then use the callbacks to log which files are using what elements.  

### Alternative Not-Very-Good-At-All Solution
Our other option to solve our validation problem is to write a bunch of small schemas to test each feature separately but this has several problems: 
1. There are numerous subfeatures, so it would be a huge amount of work 
2. It would be a maintenance nightmare, as changes to the main schema would have to be propagated each subschema 
3. We actually need to know not only whether a given file passes or fails, but whether it uses a given feature.  In other words, the result we need is three-valued, not boolean.

#### Option 1: Callback
A callback that gets triggered for a partial match, as described above.   This could be an optional feature (eg needing to be turned on explicitly in the Options structure) to avoid slowing things down in the common case when it is not used.  Then a simple table-driven program could verify and report on all the subschemas in a single pass over each JSON file.

#### Option 2: Tabulated Results
It might also be possible to structure this like errors, logging results to an array that is accessed only after the validation is complete.  However, this would not be as flexible since it's unclear what data would need to be logged.  Obviously the location of the subschema and the validity of the match needs to be logged, and maybe the matching position in the data, both identified by a JSON pointer.  In this case a simple option to log ALL matches might be simplest; then the user would have to sort through the results to find what they want.  Alternatively, a set of sub-schema locations/ids could be given in an array (eg in an option) or via a set of function calls.

### Will you be able to implement it?
Maybe... we are looking through the code base now.   We may attempt it, as modifying ajv looks to be less work than all our other options.  But ideally if we do invest the time it would be nice to see it upstreamed.

We did try using a custom keyword to trigger a callback but it does not do what we want, since those callbacks get triggered before the subschema has finished validation, not after.  It would also be nice to support this without having to add custom keywords to the schema or otherwise modifying it (although we could live with that if we had to).

**Comment by mmccool** [comment]
Anyway, this is very similar to "match variables" in regular expressions ($1, $2, and the like) that let you extract the parts of the input data that match, and would have a range of similar applications (eg to extract specific data from a JSON file, not just validate it).

**Comment by epoberezkin** [comment]
@mmccool What you want seems  related to "annotations" collection process defined in draft-07 (ajv does not implement annotations collection): https://json-schema.org/latest/json-schema-validation.html#rfc.section.3.3.

This process may be required for draft-08 support as there will be added a validation keyword that will be dependent on the properties in subschemas that passed validation: https://github.com/json-schema-org/json-schema-spec/issues/602 (looks like it was bumped to draft-09 though).

Given that draft-08 seems to be going to define some standardised output formatting as well, it probably has to be addressed together.

It seems to me that what you need is annotations collection (your option 2).

**Comment by mmccool** [comment]
Good point about the annotations, it does indeed look to be related.  The JSON spec for that also notes several complications to the implementation of this, namely dealing with "not" and failed branches of "anyOf", "oneOf", etc.   But it sounds like that to be compliant with draft-07/08 something like this will be *necessary* in ajv eventually.

I should note that my ideal solution to this would not add any additional keywords to the schema to mark "match" points, although we (eg the standard validation use case I cited) could live with that if it were necessary.

As for the API, I personally think the callback (option 1) would be more flexible, but logging (option 2) might make it easier to deal with "not" etc (post-processing of the log could be used to remove "negative" matches, perhaps).   In general even if there are call-backs, it would be OK to defer the callbacks until processing is completed (eg after processing completes, ajv would scan the match log and trigger callbacks on the matches for which callbacks were requested).  I guess upon reflection the functionality of these options can be made equivalent, so the exact structure of the API is not a critical issue.

**Comment by mmccool** [comment]
So I may have some time to work on this now.  As a first cut, I was thinking of just cloning how errors are handled when allErrors is turned on, except it should log things when a match succeeds, instead of when it fails.  This should give option 2, the logging approach, above.  Also, it seems to me that errors should also already be set up to deal with not, oneOf, etc. correctly already.

**Comment by epoberezkin** [comment]
Annotation collection is optional in draft-07 but I don't mind it added according to the spec based on some option. It’s not a small change, particularly supporting it for various combinations of options and taking short-circuiting validation into account (there is some note on that in the spec).

**Comment by mmccool** [comment]
Well, I'm not sure I need annotation collection as defined in the spec for what I'm doing.  I just need an "anti-error" log (eg a log of successful matches, rather than unsuccessful ones).  So that's what I'm going  to do... but it might be a useful first step to supporting annotation collection.

**Comment by mmccool** [comment]
Re assertions, I don’t think we need full support as defined in the JSON spec.  The observation was just that a “match callback” (or log) would be useful for implementing them.

Regarding your idea: that might work, if such a schema could be written.  Would it be possible to write a tool that could read in a schema and write out an anti-schema?  Also: there is a “not” keyword supported in the spec (and by ajv).  Would an anti-schema simply involve putting a single “not” at the very top?
Basically, your idea is potentially useful, but we need a fast way to update the anti-schema when the schema changes.

(From: Michael McCool)

On Dec 20, 2018, at 11:16, Michael McCool <notifications@github.com<mailto:notifications@github.com>> wrote:


So I may have some time to work on this now. As a first cut, I was thinking of just cloning how errors are handled when allErrors is turned on, except it should log things when a match succeeds, instead of when it fails. This should give option 2, the logging approach, above. Also, it seems to me that errors should also already be set up to deal with not, oneOf, etc. correctly already.

**Comment by epoberezkin** [comment]
Btw, you could use $comment keyword - the hook it calls will need to be passed some extra parameters to expose the current validation state (e.g. errors collected so far and part of the data the schema was applied to), but it’s an easy change.

**Comment by wabiloo** [comment]
@mmccool - did you ever get to implement your idea?  I need something very similar (ability to decorate the JSON payload with info extracted from the schema on successful match)

**Comment by angelsvirkov** [comment]
I would also love to have a feature where you could track (or apply some logic on) the passed matches.

One suggestion by @epoberezkin was to use the $comment keyword
> Btw, you could use $comment keyword - the hook it calls will need to be passed some extra parameters to expose the current validation state (e.g. errors collected so far and part of the data the schema was applied to), but it’s an easy change.
Nevertheless, this does not scale very well when you have to maintain dozens / hundreds of JSON Schema files and make sure to include the $comment keyword on multiple places inside each file.

My suggestion would be to give a default behavior in this case. For example, call the match callback on each JSON Schema part which is not a keyword. In other words, each tested property is reported as a successful or fail match. This would be useful for business cases where you have non-technical people looking at UI which shows a percentage of passing matches or tests

**Comment by ttristan** [comment]
I totally agree with @angelsvirkov and would also love to see the feature for our use case. Adding `$comment` as a custom keyword would not be optimal, since I would have to pollute all definitions with an additional keyword. I think that would bloat the schema too much.

Is there some way to hook into each test that Ajv run without adding something to the schema itself? This way I could keep track of passed tests.

**Comment by felixfbecker** [comment]
My schemas are auto generated and each type has a `title` attribute with the name of the type. I'd like to be able to use ajv to find out from a schema with `anyOf` between all these types which one matched to read its `title` attribute. Basically, from an untyped deep JSON object and a JSON schema, find out all the type names for each object.
